pkg/registry/core/service/storage/alloc.go

```diff
@@ -940,10 +938,7 @@ func shouldAllocateNodePorts(service *api.Service) bool {
 		return true
 	}
 	if service.Spec.Type == api.ServiceTypeLoadBalancer {
-		if utilfeature.DefaultFeatureGate.Enabled(features.ServiceLBNodePortControl) {
-			return *service.Spec.AllocateLoadBalancerNodePorts
-		}
-		return true
+		return *service.Spec.AllocateLoadBalancerNodePorts
```

Comment:
@thockin my memory is blur here, it is working but I'd  like to fully understand it.

shouldAllocateNodePorts() is called from updateNodePorts() or initNodePorts().

beginCreate() -> allocateCreate() -> txnAllocNodePorts() -> initNodePorts()  
beginUpdate() allocateUpdate() -> txnUpdateNodePorts() -> updateNodePorts()

Those hooks run at the beginning, is the defaulting happening before?

https://github.com/kubernetes/kubernetes/blob/fbdd0d7b4165bc5a677d45e4dc693e3260297bfa/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L366-L386

Reply:
pkg/apis/core/v1/defaults.go

That should be done before we get into `begin*()`, when we convert from versioned to internal.  Is that not what you are seeing?

Reply:
yeah, I see that it is already defaulted, but I was trying to understand where, no probem then, this is fine, is mostly curiosity at this point
Thanks

Reply:
for ref
![image](https://user-images.githubusercontent.com/6450081/146335745-0ac14d06-a1a8-4846-848f-b246615bd89e.png)